### PR TITLE
realtek: new timer driver (v4)

### PR DESCRIPTION
**Pull request by plappermaul**
Support the realtek clock driver with a helpful companion.

### CODE REVIEW

**Comment by plappermaul** [comment]
**Notice: This timer driver needs a working IRQ affinity (see PR #10573)**

**Comment by plappermaul** [comment]
@svanheule after extensive testing with the new timer driver the patches should be fine now.

The latest version takes care about following situation:

- affinity is changed
- GIMR was not updated
- next (timer) interrupt lands on the wrong CPU

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,rtl8300-timer.yaml [hunk 1]
```diff
@@ -0,0 +1,124 @@
+# SPDX-License-Identifier: GPL-2.0
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/timer/realtek,rtl8300-timer.yaml#
```
Comment (oliv3r) [inline target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,rtl8300-timer.yaml:4]: I think you can name your timer the 'otto' timer, just like we do the gpio and wdt drivers. The architecture is the otto architecture, and assuming they wrote their own VHDL for the gpio, wdt etc drivers, then this is the otto driver(s).

If they bought some IP, we should use that driver instead of course, but I somehow feel that that is less likely, we could compare registers with other drivers of course to see if there are close similarities.

```suggestion
$id: http://devicetree.org/schemas/timer/realtek,rtl-otto-timer.yaml#
```

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,rtl8300-timer.yaml [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  reg field one timer is addressed that is taken as a stable 3.125 MHz 
+  clocksource. If the Realtek clock control unit is enabled too the required
+  clocks can be referenced by 'include/dt-bindings/clock/rtl83xx-clk.h'. 
+ 
+properties:
+  compatible:
+    enum:
+      - realtek,rtl8380-timer
+      - realtek,rtl8390-timer
+      - realtek,rtl9300-timer
```
Comment (oliv3r) [inline target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,rtl8300-timer.yaml:34]: add all 'supported' chips, for the simple reason, that device tree's should list their specific ID.

e.g. in the dts I'd expect to see 'realtek,rtl9302, realtek-rtl_otto' so that it can try its specific compatible, and fallback otherwise. Having the 'generic' rtl9300 in the middle makes sense still, e.g. if we need a chip specific quirk, we can, if it's a 'generation' (rtl9300) quirck we can, everything else is handled by otto.

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,rtl8300-timer.yaml [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+      compatible = "realtek,rtl8380-timer";
+      reg = <0x3100 0x10>, <0x3110 0x10>;
+
+      interrupt-parent = <&intc>;
+      interrupts = <7 4>;
+      clocks = <&ccu CLK_LXB>;
+    };
+  - |
+    timer0: timer@3200 {
+      compatible = "realtek,rtl9300-timer";
```
Comment (oliv3r) [inline target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,rtl8300-timer.yaml:115]:
```suggestion
      compatible = "realtek,rtl9302", "realtek,rtl9300-timer", "realtek,rtl-otto-timer";
```

File: target/linux/realtek/rtl838x/config-5.10 [hunk 4]
```diff
@@ -7,8 +7,8 @@ CONFIG_BLK_DEV_RAM=y
 CONFIG_BLK_DEV_RAM_COUNT=16
 CONFIG_BLK_DEV_RAM_SIZE=4096
 # CONFIG_BMIPS_CPUFREQ is not set
-CONFIG_CEVT_R4K=y
-CONFIG_CEVT_RTL9300=y
+# CONFIG_CEVT_R4K is not set
+# CONFIG_CEVT_RTL9300 is not set
```
Comment (oliv3r) [inline target/linux/realtek/rtl838x/config-5.10:11]: I think `make kernel_menuconfig` actually deletes these.

btw, as discussed in a different PR, I think it's fair to also delete `patches-5.10/309-cevt-rtl9300-support.patch` as part of this series.

File: target/linux/realtek/rtl839x/config-5.10 [hunk 5]
```diff
@@ -7,8 +7,8 @@ CONFIG_BLK_DEV_RAM=y
 CONFIG_BLK_DEV_RAM_COUNT=16
 CONFIG_BLK_DEV_RAM_SIZE=4096
 # CONFIG_BMIPS_CPUFREQ is not set
-CONFIG_CEVT_R4K=y
-CONFIG_CEVT_RTL9300=y
+# CONFIG_CEVT_R4K is not set
+# CONFIG_CEVT_RTL9300 is not set
```
Comment (oliv3r) [inline target/linux/realtek/rtl839x/config-5.10:11]: Weird that this is in this commit as well; as I just left a comment on the previous commit :p

File: target/linux/realtek/files-5.10/drivers/clocksource/timer-rtl83xx.c [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+		to = per_cpu_ptr(&rttm_to, cpu_rollback);
+		timer_of_cleanup(to);
+	}
+
+	return -EINVAL;
+}
+
+TIMER_OF_DECLARE(rtl838x_timer, "realtek,rtl8380-timer", rttm_probe);
+TIMER_OF_DECLARE(rtl839x_timer, "realtek,rtl8390-timer", rttm_probe);
+TIMER_OF_DECLARE(rtl839x_timer, "realtek,rtl9300-timer", rttm_probe);
```
[the remark this answers is not in the document]
Comment (oliv3r) [reply target/linux/realtek/files-5.10/drivers/clocksource/timer-rtl83xx.c:292]: Though in reality, you'd want
```suggestion
TIMER_OF_DECLARE(rtl9301_timer, "realtek,rtl9300-timer", rttm_probe);
TIMER_OF_DECLARE(rtl9302_timer, "realtek,rtl9300-timer", rttm_probe);
TIMER_OF_DECLARE(rtl9303_timer, "realtek,rtl9300-timer", rttm_probe);
```
(same for the rtl8xxx of course)

File: target/linux/realtek/dts-5.10/rtl838x.dtsi [hunk 7]
```diff
@@ -152,6 +152,15 @@
 			#size-cells = <0>;
 		};
 
+		timer0: timer@3100 {
```
Comment (oliv3r) [inline target/linux/realtek/dts-5.10/rtl838x.dtsi:155]: any reason your not defining timer1-3? putting 'status = disabled' would be fine of course, but this is the main dtsi after all.

File: target/linux/realtek/dts-5.10/rtl839x.dtsi [hunk 8]
```diff
@@ -170,6 +170,15 @@
 			#size-cells = <0>;
 		};
 
+		timer0: timer@3100 {
+			compatible = "realtek,rtl8390-timer";
+			reg = <0x3100 0x10>, <0x3110 0x10>, <0x3120 0x20>;
```
Comment (oliv3r) [inline target/linux/realtek/dts-5.10/rtl839x.dtsi:175]: Hmm, I thought you had 2 address blocks per timer; so why a) is there a 3rd (or 2times 1) and b) would that not be timer1?

Same goes for the interupt below, shouldn't that be the interrupt for the second timer?

**Comment by plappermaul** [comment]
Hi Olliver,

after [this conversation](https://github.com/openwrt/openwrt/pull/10574#issuecomment-1246321047) you qualify as the second person of today to get a polite reminder what this is all about. As stated in the other post I am thankful for each and every support to get things going on. But your comments from above (except the compilation error) seem to indicate that all that goes on in the realtek target is once again for sake of commenting or sharing opinions and not to concentrate on the content. 

If you do not know what timers (or clocksources) are about in the Linux kernel then it will be best to ask in a forum and not here -  https://forum.openwrt.org/t/support-for-rtl838x-based-managed-switches/57875 could be a good place. Nevertheless I will try to explain the commit for you. Oh fine, nothing to do. It is in the DT documentation already 

```
To make use of the timers the DT needs an address/interrupt definition per CPU core to 
provide an interrupt driven timer with interval HZ. With an additional reg field one timer 
is addressed that is taken as a stable 3.125 MHz clocksource. 
```

This timer was only initiatiated because Sander "complained" that clocks are wrong after frequency changes from the clock driver. Another way of develop-by-proxy. And regarding your request to drop CEVT_RTL9300 the first commit message clearly states 

```
... the implementation follows this way: leave the RTL930X implementation as is, provide
a new driver for RTL83XX devices only, swap RTL930X driver at a later time ...
```

Looking at all the developers we have for the realtek target we can clearly state that all of them will fit into a phone booth. Missing Brigers input is clearly a loss. Getting comments as above as an alternative is no help at all and can drive developers to hang themselves in the shower.

I'm still willing to support the project in my spare time - as everyone does. But not if reviews are of a quality like Sanders or yours of the last 24 hours. If we want to talk about stuff let us do it in the forum ot the chat. And that should go beyond "hey can you test my PR?" or "but the manual states it can be an oscillator OR a crystal".

**Comment by oliv3r** [comment]
> Hi Olliver,
> 
> after [this conversation](https://github.com/openwrt/openwrt/pull/10574#issuecomment-1246321047) you qualify as the second person of today to get a polite reminder what this is all about. As stated in the other post I am thankful for each and every support to get things going on. But your comments from above (except the compilation error) seem to indicate that all that goes on in the realtek target is once again for sake of commenting or sharing opinions and not to concentrate on the content.

Mind you, this is also my/our personal spare time. And we probably all have different end-goals. For me, the long-term goal, is to have all these changes upstreamed, nice and clean. The 'worse' it is done so here (looking at our DSA/driver stack) will take many any weeks/months of work however, mostly, because we all lack the knowledge and truely understand what's going on. Magic values? Sloppy code duplication? It will cost double the effort to rectify this later. So giving the feedback now, where you (not you-you) understand everything as its being written, have key knowledge into getting it right, so that up-streaming later is much easier. 

However, short term, I agree, that getting stuff (properly working) in is key. But look at your IRQ timer4 fixes for SMP. This failed before because the previous point wasn't done, we just dumped something in, and are suffering from it now. It happens, so if you are arguing for 'getting something in now, vs something in not' I'm with you and advocated that as well. That's why I've taken birger's huge 30 patch-set, and cleaned it up, fixed some low level stuff, fixed some bugs. I want to get this into openwrt, and exposed to people.

While I hate to admit however, openwrt to me is a bit the 'experimental' area, where we write stuff (as good as we can) to learn from it, and improve it, before mainlining it. The sad truth is, we don't have the knowledge, time and manuals to get it right from the start. So we need an experimental 'training ground' if you will.


> If you do not know what timers (or clocksources) are about in the Linux kernel then it will be best to ask in a forum and not here - https://forum.openwrt.org/t/support-for-rtl838x-based-managed-switches/57875 could be a good place. Nevertheless I will try to explain the commit for you. Oh fine, nothing to do. It is in the DT documentation already

I'm reasonable familiar with clock sources and clock tree's, having battled them enough on arm :)

> 
> ```
> To make use of the timers the DT needs an address/interrupt definition per CPU core to 
> provide an interrupt driven timer with interval HZ. With an additional reg field one timer 
> is addressed that is taken as a stable 3.125 MHz clocksource. 
> ```

Right, so this is where it's confusing. I know 'generally' in linux we don't care about the general timers a SoC supports and is often 'used' for the kernel timer. Fine, no problem. In the above (which I'm still not 100% understanding what you wrote btw), is that we need 2 timers for the CPU/SMT subsystem. One timer runs at 1 Hz intervals and provides the kernel with a 1hz timer, the other timer, is a 'stable 3.125MHz clock source'. This where it gets weird and probably a little hackery? I would have expected that we define 2 timers as you stated, use one timer as a 'clk src' for a fixed clk, and timer0. So this kind of feels counter-intuitive is all.


> 
> This timer was only initiatiated because Sander "complained" that clocks are wrong after frequency changes from the clock driver. Another way of develop-by-proxy. And regarding your request to drop CEVT_RTL9300 the first commit message clearly states

Mind you, it is not so much 'develop-by-proxy', but helping each other out, and of course, scratching our own itch.

> 
> ```
> ... the implementation follows this way: leave the RTL930X implementation as is, provide
> a new driver for RTL83XX devices only, swap RTL930X driver at a later time ...
> ```

And I was in part responding to this, why not drop that now? You have nearly everything in place? Also, for me personally, I don't have RTL83xx hardware, only RTL93xx, so to be able to test, I am doing the patches locally so I can actually test your great work :)

> 
> Looking at all the developers we have for the realtek target we can clearly state that all of them will fit into a phone booth. Missing Brigers input is clearly a loss. Getting comments as above as an alternative is no help at all and can drive developers to hang themselves in the shower.

A small phone booth :) but mind you, I've been with this realtek situation for 42 days, and started small, cleaning stuff getting familiar with a lot of stuff, in those 42 days, 41 days I tried to hang myself in the shower, due to the very poor quality of the existing stuff. So we are not that dissimilar :)

> 
> I'm still willing to support the project in my spare time - as everyone does. But not if reviews are of a quality like Sanders or yours of the last 24 hours. If we want to talk about stuff let us do it in the forum ot the chat. And that should go beyond "hey can you test my PR?" or "but the manual states it can be an oscillator OR a crystal".

I'm in the chat through matrix, so have history, and have dropped more questions then I had answers for, I'm more then happy to discuss things there :) but as you said, it's a small phone booth somewhere out in an empty field :(

**Comment by plappermaul** [comment]
> 'used' for the kernel timer. Fine, no problem. In the above (which I'm still not 100% understanding what you wrote btw), is that we need 2 timers for the CPU/SMT subsystem. One timer runs at 1 Hz intervals and provides the kernel with a 1hz timer, the other timer, is a 'stable 3.125MHz clock source'. This where it gets weird and probably a little hackery? I would have expected that we

This means:

- we set NR_CPU+1 Timers to frequency of 3.125 MHz
- NR_CPU timers will kick an interrupt every HZ (= CONFIG_HZ = 100 Hz) to serve for e.g. generic sleep operations. Count-to value = 3125000/100
- Each of NR_CPUS timer has its own IRQ that is bound via affinitiy to one (VPE) core (eg IRQ 28-> CPU 0, IRQ 27 -> CPU 1)
- None of them can be used as a clocksource because their values are reset after each IRQ
- So we take one additional timer and use its counter register als clocksource

Better now?

**Comment by plappermaul** [comment]
1 VPE = 1 THREAD. so we have:

- 8380: 1 core, 1 thread: 1 timer + 1 clock
- 8390: 1 core, 2 threads: 2 timer + 1 clock
- 9300: 1 core, 2 threads: 2 timer + 1 clock
- 9310: ???

I explained the number of elements in the DT

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,rtl8300-timer.yaml (hunk 2 above)
Comment (plappermaul) [reply target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,rtl8300-timer.yaml:34]: In the clock driver we aligned to 0 (instead of x) as last digit. With our current DT structure keep this consistent with the clock driver.

File: target/linux/realtek/dts-5.10/rtl838x.dtsi (hunk 7 above)
Comment (plappermaul) [reply target/linux/realtek/dts-5.10/rtl838x.dtsi:155]: We need only one timer for the whole system (see other implementations).

**Comment by plappermaul** [comment]
@oliv3r If there is anything that is really needed (major) in this PR please tell me. Regarding the DT naming conventions please elaborate with @svanheule before I change them multiple times.

As a reminder: We put that into the root DT (e.g. rtl838x.dtsi) so we do not need to take care about the last digit.

**Comment by oliv3r** [comment]
Sander and I are contemplating we basically have 2 9300 in the 9310, e.g. 2 CPU with 2 vpe each, so 5 timers?

**Comment by oliv3r** [comment]
I think we are mostly good, I'm planning on anassive overhaul of all our DTS files to clean them up, after the storm has settled a bit with with all your amazing patches :)

On 15 September 2022 11:01:21 CEST, plappermaul ***@***.***> wrote:
***@***.*** If there is anything that is really needed (major) in this PR please tell me. Regarding the DT naming conventions please elaborate with @svanheule before I change them multiple times.

**Comment by oliv3r** [comment]
I noticed this as well, but often it is also timer_clk or timer_CPU named. So a thought for the alias of the node?

On 15 September 2022 10:58:46 CEST, plappermaul ***@***.***> wrote:
***@***.*** commented on this pull request.

**Comment by plappermaul** [comment]
> Sander and I are contemplating we basically have 2 9300 in the 9310, e.g. 2 CPU with 2 vpe each, so 5 timers?

That would be. But maybe we can use the MIPS GIC timer instead. Whatever is stable against CPU frequency changes.

**Comment by svanheule** [comment]
With the interrupt changes fixed, I think this set should come next to allow for reliable behaviour when reclocking the CPU.

Regarding the DT binding, the main thing to keep in mind is that it should be a description of the hardware, not a settings file for the Linux driver. If each timer has a dedicated IRQ line, those should be all specified in the devicetree description.

For the compatibles, I would go with a SoC-specific + fallback compatible: `"realtek,rtl8380-timer", "realtek,otto-timer"`. That way, you only have to bind to the fallback compatible for now, and can ignore the SoC-specific ones as long as we don't know of any SoC-specific quirks (or if you want to e.g. not provide a clocksource on RTL931x).

In my [proposed binding](https://lore.kernel.org/linux-devicetree/7c53821386b8f4c1c0ac440f1cd186e09f4a0456.1642369117.git.sander@svanheule.net/),  I went for a single reg per node, since the timers are just N implementations of the same IP block (or so it appears). That would simplify the binding, but would probably push the complexity into the driver (which I haven't looked at in detail). I'll try to provide some feedback on the driver in the coming week.

**Comment by plappermaul** [comment]
> In my [proposed binding](https://lore.kernel.org/linux-devicetree/7c53821386b8f4c1c0ac440f1cd186e09f4a0456.1642369117.git.sander@svanheule.net/), I went for a single reg per node

More than discouraged to go on here. So once again I must be honest: 

- adding this info 12 days ago with the PR tags would have saved a lot of @oliv3r spam
- adding this info a month ago when you tested the reclocking would have saved me days
- adding the driver 8 months ago to OpenWrt first would have been excellent

And to be even more honest. We not only have unsed code we even have two different types of unused code now. To make it not feel like development by proxy this time I would happy be to get more then nit-picks this time.

**Review by svanheule — COMMENTED** [review]
Please also run checkpatch.pl; it will tell you about the `NR_CPUS` thing too.

**Comment by plappermaul** [comment]
New version posted.

- DT is now hardware-describing and no longer driver-describing.
- NR_CPUS removed

**Comment by plappermaul** [comment]
After the SMP fix we should concentrate this. At least we some steps to go to avoid Ollivers out of tree builds.

- Integrate this
- use this for RTL930x & remove the old CEVT 9300 code
- finally activate SMP for RTL930x

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,otto-timer.yaml [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+  speed is derived from the Lexra Bus (LXB) by dividers. Each timer has a block
+  of 4 control registers in the adress range 0xb800xxxx with following start
+  offsets.
+
+  RTL838X: 0x3100, 0x3110, 0x3120, 0x3130, 0x3140
+  RTL839X: 0x3100, 0x3110, 0x3120, 0x3130, 0x3140
+  RTL930X: 0x3200, 0x3210, 0x3220, 0x3230, 0x3240, 0x3250
+  RTL931X: 0x3200, 0x3210, 0x3220, 0x3230, 0x3240, 0x3250, 0x3270
+
+  The first two timers are connected to an interrupt line. The driver takes the
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,otto-timer.yaml:23]: It took me about half a day to understand what's going on here. Spoiler: it's Realtek shenanigans again!

When any of the timers are started with interrupts enabled, wrap-around will cause the interrupt pending bit to be set in the timer's INT registers *and* a corresponding bit will be set in the SoC's interrupt controller. This follows the (known) mapping:
* TC0: IRQ 29
* TC1: IRQ 28
* TC2: IRQ 17
* TC3: IRQ 16
* TC4: IRQ 15

This matches the following definitions from the XGS1250-12 SDK:

```c
#define RTL8380_GLBL_INTR_MSK_ADDR                   (0x3000)
  #define RTL8380_GLBL_INTR_MSK_TC0_IE_OFFSET        (29)
  #define RTL8380_GLBL_INTR_MSK_TC1_IE_OFFSET        (28)
  #define RTL8380_GLBL_INTR_MSK_TC2_IE_OFFSET        (17)
  #define RTL8380_GLBL_INTR_MSK_TC3_IE_OFFSET        (16)
  #define RTL8380_GLBL_INTR_MSK_TC4_IE_OFFSET        (15)
```

However, putting one of TIMER2-TIMER4 as the first timer in the list caused hangs on boot, so that obviously didn't work (as you probably found out too).

The SDK also has the following definitions:
```c
#define RTL8380_INTR_ROUTINE_0_ADDR                  (0x3008)
  #define RTL8380_INTR_ROUTINE_0_TC0_RS_OFFSET       (20)
  #define RTL8380_INTR_ROUTINE_0_TC0_RS_MASK         (0xF << RTL8380_INTR_ROUTINE_0_TC0_RS_OFFSET)
  #define RTL8380_INTR_ROUTINE_0_TC1_RS_OFFSET       (16)
  #define RTL8380_INTR_ROUTINE_0_TC1_RS_MASK         (0xF << RTL8380_INTR_ROUTINE_0_TC1_RS_OFFSET)
```
IRR0 has routing indices 31-24. This puts the routing indices for TC0 and TC1 at 29 and 28, matching the earlier `IE_OFFSET` values.

The "fun" begins when one then looks at what the TC2-TC4 values are:
```c
#define RTL8380_INTR_ROUTINE_2_ADDR                  (0x3010)
  #define RTL8380_INTR_ROUTINE_2_TC2_RS_OFFSET       (28)
  #define RTL8380_INTR_ROUTINE_2_TC3_RS_OFFSET       (24)
  #define RTL8380_INTR_ROUTINE_2_TC4_RS_OFFSET       (20)
```
IRR2 has routing indices 15-8 (see where this is going), putting {TC2, TC3, TC4} at routing indices {15, 14, 13}. When putting this translation into the IRQ driver, TC4 wakes up the CPU as expected (TC2 and TC3 untested for now).

TL;DR: The interrupt controller is going to require a table mapping IRR offsets to IMR/ISR offsets. Thanks, Realtek.

This means you are able to specify an interrupt line for every timer. I would suggest to specify the ISR offsets as the canonical HW IRQ numbers, and keep the routing table index mapping internal to the IRQ driver. The latter can be implemented later, as TC0 and TC1 are mapped as expected, and are the only ones needed for now.

**Comment by oliv3r** [comment]
> After the SMP fix we should concentrate this. At least we some steps to go to avoid Ollivers out of tree builds.
> 
>     * Integrate this
> 
>     * use this for RTL930x & remove the old CEVT 9300 code
> 
>     * finally activate SMP for RTL930x

Well my out-of-tree build is to also integrate all the networking functionality birger did :) the timer is just a tiny thing :p but yes! absolutely! And I'll re-base your patches to test them asap!

**Comment by plappermaul** [comment]
v3 pushed

- update documentation (6th IRQ on RTL930x unknown and left out for now)
- update DTSI
- disable new config for 93xx targets

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,otto-timer.yaml (hunk 9 above)
Comment (svanheule) [reply target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,otto-timer.yaml]: Writing to the registers at 0x3250-0x325f doesn't have any effect (XGS1250-12, RTL9302B), so it appears TC5 isn't actually there. The watchdog at 0x3260 is correct though, so looks like the left a gap for TC5, but didn't actually include it.

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,otto-timer.yaml [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+  RTL838X: 0x3100, 0x3110, 0x3120, 0x3130, 0x3140
+  RTL839X: 0x3100, 0x3110, 0x3120, 0x3130, 0x3140
+  RTL930X: 0x3200, 0x3210, 0x3220, 0x3230, 0x3240, 0x3250
+  RTL931X: 0x3200, 0x3210, 0x3220, 0x3230, 0x3240, 0x3250, 0x3270
+
+  The driver takes the first n register and interrupt definitions as clock
+  events for each (virtual) CPU. The n'th+1 timer is taken as a stable 3.125 MHz
+  clocksource. If the Realtek clock control unit is enabled too the required
+  clocks can be referenced by 'include/dt-bindings/clock/rtl83xx-clk.h'.
+
```
Comment (svanheule) [inline target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,otto-timer.yaml:27]: You missed a checkpatch.pl error:
```
ERROR: trailing whitespace
#47: FILE: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,otto-timer.yaml:27:
+ $
```

**Comment by plappermaul** [comment]
trying to clean up others leftovers, always complaining, never happy, compiling for hours ...

**Comment by plappermaul** [comment]
v4 pushed

- init timer with 0 instead of 0x0fffffff (seems to work)
- change clocksource initialization
- update documentation
- change DTSI

File: target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,otto-timer.yaml (hunk 9 above)
Comment (oliv3r) [reply target/linux/realtek/files-5.10/Documentation/devicetree/bindings/realtek,otto-timer.yaml]: Since there's seems to be no IRQ for it; I'm not surprised at all. Could be as simple as miss-communication between the software/spreadsheet guys and the FPGA guys, where the later removed it to save space; but software was 'already working'.

File: target/linux/realtek/patches-5.10/302-clocksource-add-otto-driver.patch [hunk 11]
```diff
@@ -0,0 +1,42 @@
+--- a/drivers/clocksource/Kconfig
++++ b/drivers/clocksource/Kconfig
+@@ -127,6 +127,18 @@ config RDA_TIMER
+ 	help
+ 	  Enables the support for the RDA Micro timer driver.
+ 
++config OTTO_TIMER
```
Comment (oliv3r) [inline target/linux/realtek/patches-5.10/302-clocksource-add-otto-driver.patch:7]: I know it's a small annoying nitpick, but to stay consistent with other drivers, could you name it
```suggestion
+config REALTEK_OTTO_TIMER
```
like we have `REALTEK_OTTO_WDT` and `REALTEK_OTTO_SPI` etc?

**Comment by svanheule** [comment]
I've implemented @oliv3r's suggestion to rename the driver symbol, swapped the order of the compatibles, and dropped some unused MIPS dependencies. Merged to my staging tree, should land in master soon.

**Comment by plappermaul** [comment]
> I've implemented @oliv3r's suggestion to rename the driver symbol, swapped the order of the compatibles, and dropped some unused MIPS dependencies. Merged to my staging tree, should land in master soon.

@svanheule Thanks.
